### ISSUE TITLE
Field filter fails if it references a column from an aliased table

### ISSUE THREAD

**Issue by maliayas** [metabase/metabase#3324]
I'm using a filter variable in an SQL query. The query is similar to this:

````sql
select bla_bla
from my_table as mt
left join bla_bla
where {{some_date_field}}
````

`some_date_field` is a variable with type "Field filter" and it's a column from `my_table`. However this filter fails in the dashboard. After some investigation, I found that Metabase generates filter's SQL code as `my_table.some_column` and not `mt.some_column` which causes a MySQL error like "Unknown column 'my_table.some_column' in 'field list'". Apparently MySQL requires using alias if an alias is defined for that table.

It seems, in the menu where we define field filters, metabase should ask user how to refer to the table & column rather than what table and column to pick. Or this may be integrated into the syntax. E.g.

````sql
select bla_bla
from my_table as mt
left join bla_bla
where {{some_date_field:mt.my_table}}
````

### Bug Report Info

*  I'm using this browser:
   - [x] Chrome
   - [ ] Firefox
   - [ ] Safari
   - [ ] Internet Explorer / Edge
   - [ ] Other: ____________________
*  My computer's OS is:
   - [ ] OS X
   - [x] Windows
   - [ ] Linux
   - [ ] Other: ____________________
*  I'm running Metabase:
   - [ ] Locally from the JAR
   - [ ] Locally with the Mac App
   - [ ] On AWS (via Docker)
   - [ ] On Heroku
   - [x] Other: Local Docker
* *Internally*, Metabase is using this database: (This is the database Metabase stores internal information like accounts and dashboards in, *not* one you're running queries against. If you haven't done anything special to configure this, Metabase defaults to using H2 when running locally; if you're running Metabase on AWS or Heroku using the provided scripts/instructions, you're most likely using Postgres.):
   - [x] H2
   - [ ] Postgres
   - [ ] MySQL
*  My data is in a database of this type:
   - [ ] Postgres
   - [x] MySQL
   - [ ] Redshift
   - [ ] BigQuery
   - [ ] MongoDB
   - [ ] SQLite
   - [ ] Druid
   - [ ] SQL Server
   - [ ] Crate
   - [ ] Oracle
   - [ ] Other: ____________________
*  I'm using Metabase version:
   - [ ] 0.18.0
   - [ ] 0.18.1
   - [ ] 0.19.0
   - [ ] 0.19.1
   - [ ] 0.19.2
   - [x] 0.19.3
   - [ ] Other: ____________________

**Comment by salsakran** [comment]
Yes, we don't really support aliased table names in this version of filter tags. Definitely seems like a useful addition. 

I like the idea of optionally being able to specify the columns aliased name, as this would solve a number of other use cases we've found where the current approach doesn't work 100% of the time.

**Comment by maliayas** [comment]
Thanks for your comments. Btw I meant `{{some_date_field:mt.some_column}}` instead of `{{some_date_field:mt.my_table}}`.

7 Eylül 2016 20:19:17 EEST, Ali Ayas <notifications@github.com> yazdı:

**Comment by nickcarenza** [comment]
What kind of dev effort would this feature take? Is it something a newcomer to the project would tackle?

**Comment by igisho** [comment]
Dos this issue has any workaround?

**Comment by gilhs** [comment]
@igisho yeah, write SQL :/

**Comment by marcoruggine** [comment]
I'm stuck here as well and it is extremely annoying. Any plan to fix this issue?

**Comment by Tim-The-Third** [comment]
@igisho @marcoruggine workaround is to use the full table name, 
so rather than 
`Select AB.name
From Alphabeta as ab`

select Alphabeta.name
from Alphabeta

**Comment by kimroen** [comment]
Could an alternative be to add behavior where you can output the ID instead of the full part of the query? That way, you could alias all you want.

What are the different behaviors we provide by emitting the full clause at the moment? Could those be handled some other way, maybe?

**Comment by kimroen** [comment]
I worked around the issue by using a number as the variable instead of a field filter and providing a link to a different Question to look up the right ID by name.

**Comment by probaner** [comment]
Facing same issue when querying SQL server with column for aliased table. Trying to use it as field filter.

**Comment by Woodz** [comment]
I would suggest that we review the design of field filters. It doesn't seem practical to have a dropdown of every database, table and then field that you want to filter on when the majority of those are not present in the query. Identifying tables and fields used in the current query (including aliases) seems like a good way to both improve the usability of field filters as well as to address this issue. Thoughts?

**Comment by calfzhou** [comment]
it would be better not automatically include field name when using field filter. 
say, i prefer `where my_field in {{filter_value}}` instead of `where {{filter_value}}`.

**Comment by maliayas** [comment]
> @calfzhou: it would be better not automatically include field name when using field filter.

You can aldreay achive that by not selecting field filter as variable type. You have "text", "number" and "date" options which work exactly like your example. "Field filter" type on the other hand needs to know how to refer to the column because the final SQL syntax will change based on what the user selects via the filter box.

> @Woodz Thoughts?

I would not suggest parsing the query in order to identify possible tables and aliases. Asking this to the user via a textbox would be a more proper way.

> @Woodz I would suggest that we review the design of field filters.

It would be great. So here is what's on my mind.

***

As far as I see, field filter variable type is useful only for 2 cases:

1. **You want a rich date selector.** "Date" type allows you only to pick a single date, where "field filter" and then "date filter" as widget type allows you choose from many date selector types.
2. **You want autocomplete and/or multiselection** while filtering on a category column. Metabase achieves this via SQL constructs like `some_column in ("name 1", "name 2")`

Now I'll try to reimagine the UX to achieve above requirements as well as addressing this GitHub issue.

Case # 1 only needs to know "how to refer to the column". A textbox in the variable details UI would be enough for this. You can enter `my_table.some_column` or `mt.some_column` if you're using an alias. Further details are irrelevant to Metabase, such as which specific table&column is that.

Case # 2 on the other hand needs to know both "how to refer to the column" and "which specific table&column is that". This is because Metabase will retrieve values from that column in order to provide autocomplete. So this UX can be achieveable by a textbox just like above and a db/table/column picker which we already have in the current variable details UI.

So I propose to both:

1. Add a fifth option to the variable type selectbox: Text, Number, Date, __Rich Date__, Field Filter. New _Rich Date_ type will solve Case # 1 above.
2. Remove "Date Filter" widget type from "Field Filter" variable type. Because it's now Rich Date, *no longer a field filter*.

Furthermore, Date and Rich Date can even be combined into one item, since the latter is the superset of the former.

This is my 2 cents. I hope the issue gets fixed in the most simple and consistent way.

**Comment by leorochael** [comment]
Being able to refer to aliases on a Field Fielter is a requirement when doing a query with an auto-join.

For example, imagine a `Persons` table with columns for `Biological Father` and `Biological Mother` that are foreign keys into the same `Persons` table.

In a query where I'm trying to filter on columns of the father or mother, I'll have to use one or more aliiases to refer to the same table.

I think it's a must to be able to to specify the alias, either using a field on the variable form, as suggested by others here, or to have some form of syntax on the variable definition inside the SQL.

The syntax option could be something like `{{table_alias.variable_name}}`, as in:

```
SELECT
   alias1.column1 AS alias1_column1,
   alias2.column1 AS alias2_column1
FROM sometable AS alias1
INNER JOIN sometable AS alias2 ON (alias1.join_id = alias2.id)
[[ WHERE {{alias1.my_variable}} ]]
```

**Comment by lwqcz** [comment]
Guys, any plans for this feature/repair? To have no feature to use aliases in complex SQLs and widgets is really a pain. 🤕  Thanks.

**Comment by Nufan1121** [comment]
Apologies if this has been chase elsewhere, this is the **only** issue i was able to find that matches my exact and current problem. Is there any traction on this Open issue please? It seems communication has halted since '27 Aug 2018' - this would be an extremely useful feature for us. 

I can confirm the filters work perfectly as an intended in my tests sans the table alias. It would be extremely unfortunate to have to correct the duplicate column names in the structure of my tables in order to use these brilliant filter selection features.

**Comment by flamber** [comment]
Hi @Nufan1121 Not sure what you mean by traction, but this is still an open issue. Don't understand the problem with duplicate columns, since this should only be an issue for table aliases.
If you need help troubleshooting, then please use the forum: https://discourse.metabase.com/

**Comment by Nufan1121** [comment]
> Hi @Nufan1121 Not sure what you mean by traction, but this is still an open issue. Don't understand the problem with duplicate columns, since this should only be an issue for table aliases.
> If you need help troubleshooting, then please use the forum: https://discourse.metabase.com/

Hi @flamber , thank you for the quick reply. Apologies for being vague, i believe my problem is the same as this one as we have two tables with identical column names (which is what i meant by duplicate column names). This is why we must use aliases or restructure to the table entirely in order to be able to use the field type filter. 

I appreciate this i still open, i was hoping to identify if there had been any progress made on this since 27 Aug 2018? I will of course review the link for forums and see if there has been any mention/progress of this issue on there too.

**Comment by flamber** [comment]
@Nufan1121 
Any progress about issues is tracked here. The forum is for questions/troubleshooting.
But you shouldn't have any issues with aliased columns, it's an issue with aliased tables.

**Comment by runningdata** [comment]
> @igisho @marcoruggine workaround is to use the full table name,
> so rather than
> `Select AB.name From Alphabeta as ab`
> 
> `select Alphabeta.name from Alphabeta`
> 
> (in summary if you plan to use a column as a field filter just don't alias it)

cannot work for me...

**Comment by rxlky** [comment]
Yes we need this feature please. Can you suggest workaround because my case is not about joining table but using where exists. So I really need table alias, and some workaround above mentioning to use full table name cannot work for me.

**Comment by sbelak** [comment]
A workaround which depending on the DB may or may not be performant is to wrap the problematic query in a subselect and have the filter defined in the outermost select. Reusing the example from the OP:
```
select bla_bla from
  (select bla_bla, some_date_field
  from my_table as mt
  left join bla_bla)
where {{some_date_field}}
```

**Comment by rxlky** [comment]
This is also not working for me because I'm using mysql and subquery like this requires table alias too.

My query something like this

> SELECT month(order_date) as month,
> COUNT(if(class='car',order_number,NULL)) AS 'car',
> COUNT(if(class='bike',order_number,NULL)) AS 'bike',
> COUNT(order_number) AS total
> FROM my_report s
> WHERE EXISTS (SELECT 1 FROM my_report t WHERE t.order_number = s.order_number AND s.class ='bike')
> GROUP BY month(order_date)

I will need s.class = 'bike' inside where exists as query filter

**Comment by sbelak** [comment]
How about if you rewrite WHERE EXISTS as a join (I'm not very familiar with MySQL, but in Postgres both get compiled to the same operations)?

**Comment by ferarnoled** [comment]
> A workaround which depending on the DB may or may not be performant is to wrap the problematic query in a subselect and have the filter defined in the outermost select. Reusing the example from the OP:
> 
> ```
> select bla_bla from
>   (select bla_bla, some_date_field
>   from my_table as mt
>   left join bla_bla)
> where {{some_date_field}}
> ```
> 
> Alternatively you could also use the unfiltered query as a source and create a nested question using the query builder.

Another option is to create a view on your DB that does all the joins so you don't need to use an alias.

**Comment by ChristineChetty** [comment]
> I worked around the issue by using a number as the variable instead of a field filter and providing a link to a different Question to look up the right ID by name.

how did you do this??

**Comment by kimroen** [comment]
> > I worked around the issue by using a number as the variable instead of a field filter and providing a link to a different Question to look up the right ID by name.
> 
> how did you do this??

Oh, I'm sorry but I don't remember anything about this and the project doesn't exist anymore so I can't check either 😞

**Comment by johan-lejdung** [comment]
This is still not working properly right? Seems like I'll have to resort to a workaround..

Does anyone know if there is work being done on this issue?

**Comment by gingeard** [comment]
I've found worked workaround.

You can replace broken expression of field filter:
[[ AND {{SomeFieldFilter}}) ]]

by wrapped version with subquery like this:
[[ AND t.col IN (SELECT col FROM your_table WHERE {{SomeFieldFilter}}) ]]

But I'm not sure if this is efficient.

**Comment by johan-lejdung** [comment]
Well, although annoyingly the solution is to no use aliases.. 

But I agree, seems like a fairly small change in their generation code to include aliases on the column references. Their codebase seems rather structured so it shouldn't be too hard to raise a PR

**Comment by khaledwaleeed** [comment]
update ?
I wonder why such important feature request still open .

**Comment by paoliniluis** [comment]
@khaledwaleeed the reason is that we don't parse the SQL that's being entered by the user, as people use Metabase for very simple queries till extremely complex SQL's. We need to consider all those cases, then process all the aliases and afterwards inject the field filter SQL. As you can imagine, this is not something that can be done in a short amount of time

**Comment by johan-lejdung** [comment]
> @khaledwaleeed the reason is that we don't parse the SQL that's being entered by the user, as people use Metabase for very simple queries till extremely complex SQL's. We need to consider all those cases, then process all the aliases and afterwards inject the field filter SQL. As you can imagine, this is not something that can be done in a short amount of time

Another, much simpler option, would be to let the user enter the alias for the table in question. That would be more than sufficient for me at least

**Comment by JamesPoel** [comment]
Is there any progress on this issue? This issue is absolutely crippling for one of my queries.

**Comment by paoliniluis** [comment]
@JamesPoel we need to rewrite a lot of how we process queries for this to be fixed

**Comment by cwe5590** [comment]
> > @khaledwaleeed the reason is that we don't parse the SQL that's being entered by the user, as people use Metabase for very simple queries till extremely complex SQL's. We need to consider all those cases, then process all the aliases and afterwards inject the field filter SQL. As you can imagine, this is not something that can be done in a short amount of time
> 
> @paoliniluis Another, much simpler option, would be to let the user enter the alias for the table in question. That would be more than sufficient for me at least

This would be very helpful. What comes to mind is some of the dbt_utils.star function's relation_alias kwarg e.g. https://github.com/dbt-labs/dbt-utils/blob/main/macros/sql/star.sql

**Comment by Tux** [comment]
> > @khaledwaleeed the reason is that we don't parse the SQL that's being entered by the user, as people use Metabase for very simple queries till extremely complex SQL's. We need to consider all those cases, then process all the aliases and afterwards inject the field filter SQL. As you can imagine, this is not something that can be done in a short amount of time
> 
> @paoliniluis Another, much simpler option, would be to let the user enter the alias for the table in question. That would be more than sufficient for me at least

Exactly my thought. Should not be too hard to implement. Currently there is a deadlock

**Comment by pvonglehn** [comment]
As an intermediate solution a specific error message would be **really** helpful 

e.g. 'Field filters cannot be used with aliased tables'

Several of my colleagues and I have independently spend considerable time debugging this same problem as it is not at all obvious if the

**Comment by paoliniluis** [comment]
@pvonglehn we'll start to parse SQL soon, so probably that solution will come at this point

**Comment by joaofonsecapromptly** [comment]
Any news on this? The solution to use Text or Date filters does not work for a lot of use cases because for example, if you want a filter for multiple values, you need field filter. If you want a more comprehensive and interactive dashboard with filters, you need field filter. And it really hinders the use if the field filter only accepts one value at a time

**Comment by tyronearnold** [comment]
I have a use case for aliasing not covered here as far as I can tell. I need to join twice to the same table in different contexts and filter with a dropdown list on multiple values on each join (so am forced to use a field filter): 

```
select order_id
             , buyer.name as buyer_name
             , seller.name as seller_name
from orders
left join companies as buyer on order.buyer_id = buyer.id
left join companies as seller on order.seller_id = seller.id 
where true
[[and buyer.name = {{buyer_name}}]] -- this is required but does not work
[[and seller.name = {{seller_name}} -- this is required but does not work
```

So companies is referenced twice here in different contexts related to orders. Simple text filters do not provide dropdown functionality and without the ability to see and filter the list of available companies (>2600), none of the reports are useful.

**Comment by relsqui** [comment]
@tyronearnold I've been using common table expressions to work around the problem -- can you do something like this?

```sql
with
  buyer as (select id from companies [[where {{buyer_name}}]]),
  seller as (select id from companies [[where {{seller_name}}]])
select id
from orders
  inner join buyer on orders.buyer_id = buyer.id
  inner join seller on orders.seller_id = seller.id
```

**Comment by felipefava** [comment]
Any updates on this? I'm trying to use a Field Filter to provide a good UX with a date range picker in a dashboard. However, my use case involves filtering based on a derived value (i.e., not a direct column in the database).

The SQL I need to execute looks like this:
```
SELECT COUNT(*)
FROM subscriptions
WHERE COALESCE(subscriptions.cancel_at, subscriptions.canceled_at) BETWEEN date_1 AND date_2
```

The problem is:
- I can't bind a Field Filter to a custom SQL expression like COALESCE(...)
- I can't bind a Field Filter to an alias
- I also can’t map a single dashboard filter to multiple variables, so a workaround like this breaks UX:
```
SELECT COUNT(*)
FROM subscriptions
WHERE (
  (subscriptions.cancel_at IS NOT NULL AND {{cancel_at_filter}})
  OR
  (subscriptions.cancel_at IS NULL AND subscriptions.canceled_at BETWEEN {{start_date}} AND {{end_date}})
)
```

I want to let users select a single date range in the dashboard and apply that filter to a logical "cancellation date" (cancel_at if present, otherwise canceled_at), but currently there's no clean way to do this without duplicating filters or degrading the user experience.

This feels like a pretty typical use case. I'm not a Metabase expert, so maybe there’s an easy solution I don't know or is not currently possible?

Would appreciate any pointers or suggestions!

**Comment by dragonsahead** [comment]
@felipefava build a view with the coalesce function and then use the field filter on the view (or better, use a gui question)
